sling/nlp/ner/annotators.cc

```diff
@@ -0,0 +1,1062 @@
+    {"Q838948",    SPAN_ART},              // work of art
+    {nullptr, 0},
+  };
+
+  // Build taxonomy as well as mapping from types to span flags.
+  std::vector<Text> types;
+  for (auto *type = span_taxonomy; type->first != nullptr; ++type) {
+    const char *name = type->first;
+    int flags = type->second;
+    Handle t = store->LookupExisting(name);
+    if (t.IsNil()) {
+      LOG(WARNING) << "Ignoring unknown type in taxonomy: " << name;
+      continue;
+    }
+    type_flags_[t] = flags;
+    types.push_back(name);
+  }
+
+  catalog_.Init(store);
+  taxonomy_ = new Taxonomy(&catalog_, types);
+}
+
+void SpanTaxonomy::Annotate(const PhraseTable &aliases, SpanChart *chart) {
+  const Document *document = chart->document();
+  Store *store = document->store();
+  PhraseTable::MatchList matchlist;
+
+  // Run over all spans in the chart.
+  for (int b = 0; b < chart->size(); ++b) {
+    int end = std::min(b + chart->maxlen(), chart->size());
+    for (int e = b + 1; e <= end; ++e) {
+      SpanChart::Item &span = chart->item(b, e);
+
+      if (span.aux.IsRef() && !span.aux.IsNil()) {
+        // Classify span based on single resolved item.
+        Frame item(store, span.aux);
+        int flags = Classify(item);
+        span.flags |= flags;
+      } else if (span.matches != nullptr) {
+        // Get the matches for the span from the alias table.
+        aliases.GetMatches(span.matches, &matchlist);
+        CaseForm form = document->Form(b + chart->begin(), e + chart->begin());
+        bool matches = false;
+        for (const auto &match : matchlist) {
+          // Skip candidate if case forms conflict.
+          if (match.form != CASE_NONE &&
+              form != CASE_NONE &&
+              match.form != form) {
+            continue;
+          }
+
+          // Classify item.
+          Frame item(store, match.item);
+          int flags = Classify(item);
+          span.flags |= flags;
+
+          // Titles of works of art can add a lot of spurious matches to the
+          // chart because almost any word or short phrase can be the title
+          // of a song, book, painting, etc. These are only included if the
+          // length of the matching phrase is above a threshold and not all
+          // lower case.
+          if (flags & SPAN_ART) {
+            if (form != CASE_LOWER && e - b > min_art_length) matches = true;
+          } else {
+            matches = true;
+          }
+        }
+
+        // Discard matches if all candidates have conflicts.
+        if (!matches) {
+          span.matches = nullptr;
+        }
+      }
+    }
+  }
+}
+
+int SpanTaxonomy::Classify(const Frame &item) {
+  Handle type = taxonomy_->Classify(item);
+  if (type.IsNil()) return 0;
+  auto f = type_flags_.find(type);
+  if (f == type_flags_.end()) return 0;
+  return f->second;
+}
+
+std::unordered_set<string> PersonNameAnnotator::particles = {
+  "de", "du", "di", "von", "van"
```

Comment:
This list only contain "name prepositions" so far, whereas "der" is a determiner". We could add more of these, but I have kept this list short so far because it is not language-dependent at the moment, e.g. "der" means there/that in Danish. I might have to make this list (as well as other short word lists) in the annotators language dependent. Then we can more easily add more words without introducing too much ambiguity.